File: src/wings/cli/logging.py
"""Request/response logger — writes session transcripts to .wings/logs/."""

from __future__ import annotations

import hashlib
import json
import time
from datetime import datetime, timezone
from pathlib import Path
from typing import Any


class TurnLogger:
    """Captures each API call cycle (request + response) with tool results.

    When enabled via --log, writes one .log file per session to
    .wings/logs/ in the working directory.
    """

    def __init__(self, working_dir: Path):
        self._dir = working_dir / ".wings" / "logs"
        self._dir.mkdir(parents=True, exist_ok=True)
        self._path = self._make_path()
        self._buffer: list[dict[str, Any]] = []
        self._cycle_count = 0
        self._session_start = time.monotonic()

    def _make_path(self) -> Path:
        ts = datetime.now(timezone.utc).strftime("%Y-%m-%d_%H-%M-%S")
        raw = f"{ts}-{id(self)}".encode()
        h = hashlib.sha256(raw).hexdigest()[:8]
        return self._dir / f"{ts}_{h}.log"

    @property
    def path(self) -> Path:
        return self._path

    def record_cycle(
        self,
        *,
        model: str,
        context: str = "main",
        message_count: int = 0,
        input_summary: str = "",
        response: dict[str, Any],
        system_prompt: str = "",
        tool_calls: list[str] | None = None,
        tool_results: list[str] | None = None,
        thinking: str | None = None,
    ) -> None:
        """Record a single API call cycle.

        Args:
            model: api_id used for this call (e.g. "anthropic/claude-opus-4-6")
            context: task_type — "main", "subagent/explore", etc.
            message_count: total messages in conversation (tracks context growth)
            input_summary: what was sent to the model this cycle (user prompt
                or tool result summary)
            response: the model's response (text blocks + tool use blocks)
            tool_calls: list of tool names called in this cycle
            tool_results: truncated tool outputs from this cycle
            thinking: thinking/reasoning text if the model produced it
        """
        self._cycle_count += 1
        provider_name, _, service_model = model.partition("/")
        entry: dict[str, Any] = {
            "cycle": self._cycle_count,
            "context": context,
            "timestamp": datetime.now(timezone.utc).isoformat(),
            "elapsed_s": round(time.monotonic() - self._session_start, 3),
            "provider": provider_name,
            "service_model": service_model or model,
            "api_id": model,
            "message_count": message_count,
            "input": input_summary,
            "response": response,
            "tool_calls": tool_calls or [],
        }
        if system_prompt:
            entry["system_prompt"] = system_prompt
        if tool_results:
            entry["tool_results"] = tool_results
        if thinking:
            entry["thinking"] = thinking
        self._buffer.append(entry)
        self._flush()

    def _flush(self) -> None:
        """Write buffer to the log file incrementally."""
        with open(self._path, "w") as f:
            for entry in self._buffer:
                f.write(json.dumps(entry, ensure_ascii=False, default=str))
                f.write("\n")

File: src/cli/logging.ts
/**
 * Request/response logger — writes session transcripts to .wings/logs/.
 *
 * Ported from src/wings/cli/logging.py.
 */

import { createHash } from "node:crypto";
import { mkdirSync, writeFileSync } from "node:fs";
import { join } from "node:path";

export class TurnLogger {
  private _dir: string;
  private _path: string;
  private _buffer: Array<Record<string, unknown>> = [];
  private _cycleCount = 0;
  private _sessionStart: number;

  constructor(workingDir: string) {
    this._dir = join(workingDir, ".wings", "logs");
    mkdirSync(this._dir, { recursive: true });
    this._path = this._makePath();
    this._sessionStart = Date.now();
  }

  private _makePath(): string {
    const now = new Date();
    const ts = now.toISOString().replace(/[T:]/g, "-").replace(/\..+/, "");
    const raw = `${ts}-${Date.now()}`;
    const h = createHash("sha256").update(raw).digest("hex").slice(0, 8);
    return join(this._dir, `${ts}_${h}.log`);
  }

  get path(): string {
    return this._path;
  }

  recordCycle(opts: {
    model: string;
    context?: string;
    message_count?: number;
    input_summary?: string;
    response: Record<string, unknown>;
    system_prompt?: string;
    tool_calls?: string[];
    tool_results?: string[];
    thinking?: string | null;
  }): void {
    this._cycleCount++;
    const [providerName, , serviceModel] = opts.model.split("/");
    const elapsed = ((Date.now() - this._sessionStart) / 1000).toFixed(3);

    const entry: Record<string, unknown> = {
      cycle: this._cycleCount,
      context: opts.context ?? "main",
      timestamp: new Date().toISOString(),
      elapsed_s: elapsed,
      provider: providerName ?? "",
      service_model: serviceModel ?? opts.model,
      api_id: opts.model,
      message_count: opts.message_count ?? 0,
      input: opts.input_summary ?? "",
      response: opts.response,
      tool_calls: opts.tool_calls ?? [],
    };
    if (opts.system_prompt) entry["system_prompt"] = opts.system_prompt;
    if (opts.tool_results) entry["tool_results"] = opts.tool_results;
    if (opts.thinking) entry["thinking"] = opts.thinking;

    this._buffer.push(entry);
    this._flush();
  }

  private _flush(): void {
    const lines = this._buffer.map((e) => JSON.stringify(e));
    writeFileSync(this._path, lines.join("\n") + "\n");
  }
}
